crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,377 @@
+                    )]));
+                }
+
+                diagnostics.push(invalid_capitalization);
+            }
+        }
+
+        // Check the rest of the capture groups for errors.
+        for capture_group_index in 2..=NUM_CAPTURE_GROUPS {
+            if captures.get(capture_group_index).is_some() {
+                continue;
+            }
+
+            if let Some(diagnostic) =
+                get_regex_error(capture_group_index, *token_range, &mut diagnostics)
+            {
+                diagnostics.push(diagnostic);
+            };
+        }
+
+        // If we've gotten all the way here, we know that the current token is a TODO. TDO003
+        // requires that the next token is an issue link.
+        let todo_start = find_first_t_position(comment);
+        if let Some((next_token, _next_range)) = iter.peek() {
+            if let Tok::Comment(next_comment) = next_token {
+                if ISSUE_LINK_REGEX_SET.is_match(next_comment) {
+                    continue;
+                }
+            }
+
+            diagnostics.push(Diagnostic::new(
+                MissingLinkInTodo,
+                TextRange::new(
+                    token_range.start() + TextSize::new(todo_start),
+                    token_range.start() + TextSize::new(todo_start) + TextSize::new(TODO_LENGTH),
+                ),
+            ));
+        } else {
+            // There's a TODO on the last line of the file which can't have a link after it.
+            diagnostics.push(Diagnostic::new(
+                MissingLinkInTodo,
+                TextRange::new(
+                    token_range.start() + TextSize::new(todo_start),
+                    token_range.start() + TextSize::new(todo_start) + TextSize::new(TODO_LENGTH),
+                ),
+            ));
+        }
+    }
+
+    diagnostics
+}
+
+/// Mapper for static regex errors caused by a capture group at index i (i > 1 since the tag
+/// capture group could lead to multiple diagnostics being pushed)
+fn get_regex_error(
+    i: usize,
+    range: TextRange,
+    diagnostics: &mut [Diagnostic],
+) -> Option<Diagnostic> {
+    match i {
+        2usize => Some(Diagnostic::new(MissingAuthorInTodo, range)),
+        3usize => Some(Diagnostic::new(MissingColonInTodo, range)),
+        4usize => {
+            if diagnostics
+                .last()
+                .map_or(true, |last| last.kind != MissingColonInTodo.into())
+            {
+                Some(Diagnostic::new(MissingSpaceAfterColonInTodo, range))
+            } else {
+                None
+            }
+        }
+        5usize => Some(Diagnostic::new(MissingTextInTodo, range)),
+        _ => None,
+    }
+}
+
+fn get_captured_matches(text: &str) -> CaptureMatches {
+    TODO_REGEX.captures_iter(text)
+}
+
+fn find_first_t_position(comment: &str) -> u32 {
+    // We know that the comment has to have a "t" or "T" in it, so the unwrap is safe.
+    comment.find(['t', 'T']).unwrap().try_into().unwrap()
```

Comment:
@MichaReiser Apologies if this is a basic question. Is there a better way to handle the second `unwrap()` call? This is my first PR really handling the new `TextRange/TextSize` structs, so not sure if there's a more idiomatic way.

I'm also curious - how do we handle `u32` overflow (for example, if a file is really long?) with the two new structs?

Reply:
> Apologies if this is a basic question. Is there a better way to handle the second unwrap() call?

You can do:

```
comment.find(['t', 'T']).and_then(|pos| TextSize::try_from(pos).ok()).unwrap()
```

I recommend you changing the function's return type to `TextSize`. Using `TextSize` communicates that this is a byte offset position and not any other index (or integer). 



> I'm also curious - why do those two structs use u32 over usize as their underlying offset measure?

The motivation is that `u32` requires halve the space compared to an `usize` on 64 bit systems and that should be more than sufficient for any reasonable program (you can represent files up to 4 GB). The result is that ruff uses less memory and is faster because it reads and writes less data, and more data fits into a single L1 cache line.

Reply:
Ah makes sense. Thanks for the tip and detailed explanation!